### PR TITLE
[refactor] Delete AssetGroup

### PR DESCRIPTION
**Pull request by smackesey**
## Summary & Motivation

Delete `AssetGroup` (in `dagster._legacy`).

From an examination of the code, the only place `AssetGroup` was used in an internal pathway was during workspace construction. When examining a Python module, if no job/pipeline/definitions/repo was found, then all assets would be scraped into a container `AssetGroup`. This just served to bundle the assets for a short time, until they could be dumped into a list of repository definitions when converting the loadable target to a repository.

This PR dispenses with the container `AssetGroup` and just passes a list containing all of the `AssetsDefinition` and `SourceAsset` objects in its place. This required only a few adjustments, since none of the functionality of `AssetGroup` is used anyway.

Aside from these functional changes, this also required updating some toys and tests to use alternatives to `AssetGroup`, like `Definitions`.

## How I Tested These Changes

Existing test suite.

### CODE REVIEW

**Comment by smackesey** [comment]
Current dependencies on/for this PR:
* master
  * **PR #13024** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/13024" target="_blank">Graphite</a> 
    * **PR #13025** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/13025" target="_blank">Graphite</a> 
      * **PR #13028** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/13028" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/13028?utm_source=stack-comment).

**Review by sryza — CHANGES_REQUESTED** [review]
Very in favor of deleting AssetGroup, but I think many of the tests in test_asset_group, etc. are the only tests that cover some important code paths.

I've been chipping away at migrating these tests elsewhere.  E.g. I merged this: https://github.com/dagster-io/dagster/pull/12606.

And I've been working on a PR that migrates most of the rest of them: https://github.com/dagster-io/dagster/pull/13033.

**Comment by smackesey** [comment]
>And I've been working on a PR that migrates most of the rest of them

Nice-- how should we proceed then? Just leave this open until you migrate them all in that PR? Or would it be easier to merge that PR as-is and migrate the rest of them in this one?

**Comment by smackesey** [comment]
@sryza Any more progress on that other PR or can it be merged and the rest of the tests ported in this one? I'm thinking this would be nice to merge together with the `PipelineDefinition/`ModeDefinition`/`PresetDefinition` deletion.

**Comment by smackesey** [comment]
>I think the path forward requires handling all the remaining tests in test_asset_group.py - i.e. by migrating them to not use AssetGroup or by confirming that they're handled by existing tests.

OK-- what if you you remove the few (small) modifications to non-test-code in the other PR and land the test port that you have, then I'll just port the rest of them here-- it looks like there's only a few left.

File: python_modules/dagster/dagster_tests/asset_defs_tests/test_assets_job.py [hunk 1]
```diff
@@ -1243,13 +1259,16 @@ def unconnected():
     pass
 
 
-asset_group = AssetGroup([foo, ab, bar, foo_bar, baz, unconnected])
+asset_defs = [foo, ab, bar, foo_bar, baz, unconnected]
+# asset_definitions = Definitions(
```
Comment (sryza) [inline python_modules/dagster/dagster_tests/asset_defs_tests/test_assets_job.py:1263]: stray comment

File: python_modules/dagster/dagster_tests/asset_defs_tests/test_assets_job.py [hunk 2]
```diff
@@ -893,7 +901,15 @@ def my_graph():
         graph_def=my_graph, partitions_def=StaticPartitionsDefinition(["a", "b", "c"])
     )
 
-    AssetGroup([assets_def]).build_job("abc").execute_in_process(partition_key="a")
+    assert (
```
Comment (sryza) [inline python_modules/dagster/dagster_tests/asset_defs_tests/test_assets_job.py:904]: I think it's better to use `materialize_to_memory` here and in similar situations

File: python_modules/dagster/dagster_tests/asset_defs_tests/test_assets_job.py (hunk 1 above)
Comment (smackesey) [reply python_modules/dagster/dagster_tests/asset_defs_tests/test_assets_job.py:1263]: good catch

File: python_modules/dagster-graphql/dagster_graphql_tests/graphql/test_assets.py [hunk 3]
```diff
@@ -1459,18 +1459,18 @@ def test_execute_job_subset(self, graphql_context: WorkspaceRequestContext):
...
-            asset_selection=[{"path": ["foo"]}, {"path": ["foo_bar"]}],
+            asset_selection=[{"path": ["foo"]}, {"path": ["bar"]}, {"path": ["foo_bar"]}],
         )
         run = graphql_context.instance.get_run_by_id(run_id)
         assert run and run.is_finished
         events = _get_sorted_materialization_events(graphql_context, run_id)
         assert len(events) == 2
-        assert events[0].get_dagster_event().asset_key == AssetKey("foo")
-        assert events[1].get_dagster_event().asset_key == AssetKey("foo_bar")
+        # assert events[0].get_dagster_event().asset_key == AssetKey("foo")
```
Comment (sryza) [inline python_modules/dagster-graphql/dagster_graphql_tests/graphql/test_assets.py:1472]: Comments

Comment (smackesey) [reply python_modules/dagster-graphql/dagster_graphql_tests/graphql/test_assets.py]: fixed

**Review by sryza — APPROVED** [review]

**Review by sryza — COMMENTED** [review]
I noticed that this does not remove asset_group.py. Should it?

**Pull request merged**